src/tools/tidy/src/rustdoc_html_ids.rs

```diff
@@ -0,0 +1,143 @@
+//! Checks that all error codes have at least one test to prevent having error
+//! codes that are silently not thrown by the compiler anymore.
+
+use std::collections::HashMap;
+use std::ffi::OsStr;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+use std::path::Path;
+
+use regex::Regex;
+
+const ID_MAP_PATH: &str = "librustdoc/html/markdown.rs";
+const IDS_TO_IGNORE: &[&str] = &[
+    // This one is created in the JS cannot be found in rust files.
+    "help",
+];
+
+fn extract_ids(path: &Path, bad: &mut bool) -> HashMap<String, usize> {
+    let file = File::open(path).expect("failed to open file to extract rustdoc IDs");
+    let buf_reader = BufReader::new(file);
+    let mut iter = buf_reader.lines();
+    let mut ids = HashMap::new();
+
+    while let Some(Ok(line)) = iter.next() {
+        if line.starts_with("fn init_id_map() -> ") {
+            break;
+        }
+    }
+    // We're now in the function body, time to retrieve the IDs!
+    while let Some(Ok(line)) = iter.next() {
+        if line.trim_start().starts_with("map.insert(\"") {
+            let id = line.split('"').skip(1).next().unwrap();
+            if ids.insert(id.to_owned(), 0).is_some() {
+                eprintln!(
+                    "=> ID `{}` is defined more than once in the ID map in file `{}`",
+                    id, ID_MAP_PATH
+                );
+                *bad = true;
+            }
+        } else if line == "}" {
+            // We reached the end of the function.
+            break;
+        }
```

Comment:
Instead of parsing rust source code wouldn't it be better to just load the list of ids from a file? `markdown.rs` could populate the map from the same file.

Reply:
That solves half the problem but that could be something we could use.

Reply:
I agree that parsing Rust code like this seems like a bad idea. I don't think we need a separate file, but having a simple list inside a macro call is going to be much easier to parse here and is more clearly a DSL that needs a specific grammar on the input (rather than arbitrary Rust code). Comments on the macro definition and invocation noting that they are parsed by tidy would be useful, too.

Reply:
I decided to move the declaration of HTML IDs into a macro so that half the problem is solved.